lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -959,6 +991,85 @@ parseAugmentedType(ApplyState &state, DictionaryAttr augmentedType,
+    // 3. The tap is on something inside a module.
+    //
+    // The case of (1) is handled WITHOUT a tap and the ground type stays on the
+    // port.  The cases of (2) and (3) are handled the same: a tap is generated,
+    // the tap is driven by the ground type, and any annotations are placed on
+    // the tap.  (2) requires port-specific handling logic.  (Also, remember
+    // that it is okay to have an output port on the RHS of a connect.  This
+    // avoids having to special case input and output ports.)
+    //
+    // TODO: Consider a dedicated tap operation instead of a dead wire tap.
+    AnnoPathValue path =
+        resolvePath(target, state.circuit, state.symTbl).getValue();
+    Operation *op = path.ref.getOp();
+    ImplicitLocOpBuilder builder(UnknownLoc::get(op->getContext()), op);
+
+    auto newTarget =
+        TypeSwitch<Operation *, TokenAnnoTarget>(op)
+            // Case (1): ground type is an external module port.
+            .Case<FExtModuleOp>([&](FExtModuleOp extModule) {
+              return tokenizePath(target).getValue();
+            })
+            // Case (2): ground type is a module port.
+            .Case<FModuleOp>([&](FModuleOp module) {
+              Value src = module.getArgument(path.ref.getImpl().getPortNo());
+              auto fieldIdx = path.fieldIdx;
+              builder.setInsertionPointToStart(module.getBody());
+              WireOp tap = builder.create<WireOp>(
+                  builder.getUnknownLoc(),
+                  path.ref.getType()
+                      .getFinalTypeByFieldID(fieldIdx)
+                      .getPassiveType(),
+                  state.getNamespace(module).newName("_gctTap"));
+              src = getValueByFieldID(builder, src, fieldIdx);
+              connectTap(builder, tap, src);
+              auto newTarget = tokenizePath(target).getValue();
+              newTarget.name = tap.name();
+              return newTarget;
+            })
+            // Case (3): ground type is inside a module.
+            .Default([&](Operation *op) {
+              auto module = op->getParentOfType<FModuleOp>();
+              Value src = op->getResult(0);
+              auto fieldIdx = path.fieldIdx;
+              builder.setInsertionPoint(op);
+              WireOp tap = builder.create<WireOp>(
+                  builder.getUnknownLoc(),
+                  path.ref.getType()
+                      .getFinalTypeByFieldID(fieldIdx)
+                      .getPassiveType(),
+                  state.getNamespace(module).newName("_gctTap"));
+              builder.setInsertionPointAfter(op);
+              src = getValueByFieldID(builder, src, fieldIdx);
+              builder.setInsertionPointAfterValue(src);
+              connectTap(builder, tap, src);
+              auto newTarget = tokenizePath(target).getValue();
+              newTarget.name = tap.name();
+              return newTarget;
+            });
+
+    SmallString<32> newTargetStr("~");
```

Comment:
I feel a deep sadness fill my soul.

Reply:
(you can use NLAs instead of firrtl path specifiers.  Ideally after scattering we would never need to parse another path specifier again)